### PR TITLE
Issue#1100 Provide argument in AutoSklearnClassifier and AutoSklearnRegressor to pass callback function

### PR DESCRIPTION
**Pull request by bitsbuffer**
This PR has implemented the functionality as described in https://github.com/automl/auto-sklearn/issues/1100
which provides an argument "get_trials_callback" that accepts a callback object which will be executed after finishing a trial.

Test case to test this functionality is added at test/test_util/test_trials_callback.py

### CODE REVIEW

**Comment by bitsbuffer** [comment]
Checkout the branch bitsbuffer:issue_1100_trials_callback from master instead development.

File: test/test_util/test_trials_callback.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        with open(os.path.join(self.output_dir, "trials.csv"), "a+") as fp:
+            fp.write(f"\n {self.trials_num}, {trial_start_time}, {trial_end_time}, {trial_status}, "
+                     f"{train_loss}, {valid_loss}, {test_loss}, {trial_classifier}")
+        self.trials_num += 1
+
+
+class VerifyTrialsCallBack(unittest.TestCase):
+
+    def test_trials_callback_execution(self):
+        base_dir = "./tmp"
```
Comment (mfeurer) [inline test/test_util/test_trials_callback.py:48]: Could you please use python's built-in `tempdir` module?

File: test/test_util/test_trials_callback.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        tmp_folder = os.path.join(base_dir, "temp")
+        output_folder = os.path.join(base_dir, "output")
+        os.makedirs("./tmp")
+        X_train, Y_train, X_test, Y_test = putil.get_dataset('breast_cancer')
+        cls = AutoSklearnClassifier(time_left_for_this_task=30,
+                                   initial_configurations_via_metalearning=0,
+                                   per_run_time_limit=10,
+                                   memory_limit=1024,
+                                   tmp_folder=tmp_folder,
```
Comment (mfeurer) [inline test/test_util/test_trials_callback.py:57]: You don't need to provide a `tmp_folder` and an `output_folder` for the unit tests.

File: test/test_util/test_trials_callback.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                                   delete_output_folder_after_terminate=False,
+                                   n_jobs=1,
+                                   include_estimators=["sgd"],
+                                   include_preprocessors=["no_preprocessing"],
+                                   get_trials_callback=AutoMLTrialsCallBack(base_dir)
+                                   )
+        cls.fit(X_train, Y_train, X_test,Y_test)
+        trials = pd.read_csv(os.path.join(base_dir, "trials.csv"))
+        assert trials.shape[0] > 1
+        assert len(os.listdir(output_folder)) > 0
```
Comment (mfeurer) [inline test/test_util/test_trials_callback.py:69]: What's in the `output_folder`?

File: test/test_util/test_trials_callback.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        os.makedirs("./tmp")
+        X_train, Y_train, X_test, Y_test = putil.get_dataset('breast_cancer')
+        cls = AutoSklearnClassifier(time_left_for_this_task=30,
+                                   initial_configurations_via_metalearning=0,
+                                   per_run_time_limit=10,
+                                   memory_limit=1024,
+                                   tmp_folder=tmp_folder,
+                                   delete_tmp_folder_after_terminate=False,
+                                   output_folder=output_folder,
+                                   delete_output_folder_after_terminate=False,
```
Comment (mfeurer) [inline test/test_util/test_trials_callback.py:60]: Could you please explain why you would like to keep the tmp and the output folder?

File: test/test_util/test_trials_callback.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                                   include_preprocessors=["no_preprocessing"],
+                                   get_trials_callback=AutoMLTrialsCallBack(base_dir)
+                                   )
+        cls.fit(X_train, Y_train, X_test,Y_test)
+        trials = pd.read_csv(os.path.join(base_dir, "trials.csv"))
+        assert trials.shape[0] > 1
+        assert len(os.listdir(output_folder)) > 0
+        shutil.rmtree(base_dir)
+
+    def test_no_trials_callback(self):
```
Comment (mfeurer) [inline test/test_util/test_trials_callback.py:72]: Could you please apply the same comments as the ones I made to the above unit test?

File: doc/index.rst [hunk 6]
```diff
@@ -96,7 +96,7 @@ reference to the following paper:
 If you are using Auto-sklearn 2.0, please also cite
 
 
- Auto-Sklearn 2.0: The Next Generation, Feurer *et al.*, to appear (2020).
+ `Auto-Sklearn 2.0: The Next Generation <https://arxiv.org/abs/2007.04074>`_, Feurer *et al.*, (arXiv, 2020).
```
Comment (mfeurer) [inline doc/index.rst:99]: Thanks for fixing this!

File: autosklearn/estimators.py [hunk 7]
```diff
@@ -52,6 +54,7 @@ def __init__(
         metric=None,
         scoring_functions: Optional[List[Scorer]] = None,
         load_models: bool = True,
+        get_trials_callback: Optional[IncorporateRunResultCallback] = None
```
Comment (mfeurer) [inline autosklearn/estimators.py:57]: Could you please add the callback to the docstring so that it's rendered in the docs?

**Review by mfeurer — COMMENTED** [review]
Thanks a lot for the contribution and please excuse the long delay. I left some notes in the comments, could you please update your code on them?

I was also wondering whether it would make sense to already allow for other callbacks, too, in case we add them to SMAC, what do you think?

File: test/test_util/test_trials_callback.py (hunk 1 above)
Comment (bitsbuffer) [reply test/test_util/test_trials_callback.py]: Hi @mfeurer, it makes sense to allow for newly added callbacks in SMAC. Modified the code to support it.

File: test/test_util/test_trials_callback.py (hunk 2 above)
Comment (bitsbuffer) [reply test/test_util/test_trials_callback.py]: modified the code to use tempfile module from Python.

File: test/test_util/test_trials_callback.py (hunk 3 above)
Comment (bitsbuffer) [reply test/test_util/test_trials_callback.py]: I was actually checking if AutoSklearnClassifier produces any output as for each successful run it creates a run directory provided ensemble_size = 0. Removed that statement as it was not required in test case. The main purpose of the callback was to gather summary of trials irrespective of trial's run status.

File: test/test_util/test_trials_callback.py (hunk 4 above)
Comment (bitsbuffer) [reply test/test_util/test_trials_callback.py]: Modified the code as it is not required.

File: test/test_util/test_trials_callback.py (hunk 5 above)
Comment (bitsbuffer) [reply test/test_util/test_trials_callback.py]: Modified the code as it is not required.

File: autosklearn/estimators.py (hunk 7 above)
Comment (bitsbuffer) [reply autosklearn/estimators.py]: Modified the code to add docstring.

**Review by mfeurer — APPROVED** [review]

**Comment by mfeurer** [comment]
Hey @bitsbuffer, please excuse the horrible delay. I have just approved the PR and made two small changes. However, there have been updates since the PR which now break your changes. Could you please rebase and update your PR, merging then will be happen quickly.

**Comment by bitsbuffer** [comment]
@mfeurer Sure, I will take care of it.

**Comment by bitsbuffer** [comment]
@mfeurer I've modified pull request which include rebase to latest development branch, also modified the test case to remove delete_output_folder_after_terminate=False.

**Comment by mfeurer** [comment]
Thanks a lot @bitsbuffer, unfortunately, the new test isn't passing and there is an issue with the automatic style checker. Could you please have a look into these?

**Comment by bitsbuffer** [comment]
@mfeurer Apologies for this. I have modified test_check_askl2_same_arguments_as_askl and code style as indicated by automatic style checker.

**Comment by mfeurer** [comment]
Hey @bitsbuffer thanks a lot for the update. Unfortunately, there are still two complaints about unnecessary imports. Could you please have a look at them?

**Comment by bitsbuffer** [comment]
Hi @mfeurer, probably I need to be more careful with this from the next time. I removed unused imports from estimator.py

**Comment by mfeurer** [comment]
Thanks a lot for your contribution!

(and yes, we need to make the process of contributing much smoother)

**Pull request merged**